pkg/util/kubernetes/hpa/hpa.go

```diff
@@ -0,0 +1,361 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"encoding/json"
+	"reflect"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	as "github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver/leaderelection"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+//// Covered by test/integration/util/kube_apiserver/hpa_test.go
+
+var (
+	expectedHPAType = reflect.TypeOf(v2beta1.HorizontalPodAutoscaler{})
+)
+
+const (
+	datadogHPAConfigMap = "datadog-hpa"
+)
+
+type CustomExternalMetric struct {
+	Name      string            `json:"name"`
+	Labels    map[string]string `json:"labels"`
+	Timestamp int64             `json:"ts"`
+	HpaOrigin string            `json:"origin"`
+	Value     int64             `json:"value"`
+}
+
+// HPAWatcherClient embeds the API Server client and the configuration to refresh metrics from Datadog and watch the HPA Objects' activities
+type HPAWatcherClient struct {
+	apiClient         *as.APIClient
+	hpaReadTimeout    time.Duration
+	hpaRefreshItl     time.Duration
+	hpaPollItl        time.Duration
+	hpaExternalMaxAge time.Duration
+}
+
+func (c *HPAWatcherClient) hpaWatcher(res string) (new []*v2beta1.HorizontalPodAutoscaler, modified []*v2beta1.HorizontalPodAutoscaler, deleted []*v2beta1.HorizontalPodAutoscaler, resVer string, err error) {
+	hpaInterface := c.apiClient.ClientSet.AutoscalingV2beta1()
+	metaOptions := metav1.ListOptions{Watch: true, ResourceVersion: res}
+
+	hpaWatcher, err := hpaInterface.HorizontalPodAutoscalers(metav1.NamespaceAll).Watch(metaOptions)
+	if err != nil {
+		log.Infof("Failed to watch %v: %v", expectedHPAType, err)
+	}
+	defer hpaWatcher.Stop()
+
+	watcherTimeout := time.NewTimer(c.hpaReadTimeout)
+	for {
+		select {
+		case rcvdHPA, ok := <-hpaWatcher.ResultChan():
+			if !ok {
+				log.Debugf("Unexpected watch close")
+				return nil, nil, nil, "0", err
+			}
+			currHPA, ok := rcvdHPA.Object.(*v2beta1.HorizontalPodAutoscaler)
+			if !ok {
+				log.Infof("Wrong type: %s", currHPA)
+				continue
+			}
+			if currHPA.ResourceVersion != "" && currHPA.ResourceVersion != resVer {
+				resVer = currHPA.ResourceVersion
+			}
+			if rcvdHPA.Type == watch.Error {
+				log.Infof("Error in the watcher, evaluating: %s", currHPA)
+				//status, ok := rcvdHPA.Object.()
+			}
+			if rcvdHPA.Type == watch.Added {
+				log.Infof("Adding this manifest: %s", currHPA)
+				new = append(new, currHPA)
+			}
+			if rcvdHPA.Type == watch.Modified {
+				log.Infof("Modifying this manifest: %s", currHPA)
+				modified = append(modified, currHPA)
+			}
+			if rcvdHPA.Type == watch.Deleted {
+				deleted = append(deleted, currHPA)
+			}
+
+			watcherTimeout.Reset(c.hpaReadTimeout)
+		case <-watcherTimeout.C:
+			return new, modified, deleted, resVer, nil
+		}
+	}
+}
+
+func newHPAWatcher() *HPAWatcherClient {
+	clientAPI, err := as.GetAPIClient()
```

Comment:
The `HPAWatcherClient` only needs a `kubernetes.Clientset` instance to be able to read and update ConfigMaps. I think we should avoid passing the `APIClient` here since we don't need any of the functionality it provides.

I think we should pass an instance of `kubernetes.Clientset` when instantiating `HPAWatcherClient`.

The singleton pattern makes it challenging to write unit tests.